lib/protocol/bufferpool.go

```diff
@@ -2,56 +2,72 @@
 
 package protocol
 
-import "sync"
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+)
 
 // Global pool to get buffers from. Requires Blocksizes to be initialised,
 // therefore it is initialized in the same init() as BlockSizes
 var BufferPool bufferPool
 
 type bufferPool struct {
-	pools []sync.Pool
+	puts   int64
+	skips  int64
+	misses int64
+	pools  []sync.Pool
+	hits   []int64 // start of slice allocation is always aligned
 }
 
 func newBufferPool() bufferPool {
-	return bufferPool{make([]sync.Pool, len(BlockSizes))}
+	return bufferPool{
+		pools: make([]sync.Pool, len(BlockSizes)),
+		hits:  make([]int64, len(BlockSizes)),
+	}
 }
 
 func (p *bufferPool) Get(size int) []byte {
 	// Too big, isn't pooled
 	if size > MaxBlockSize {
+		atomic.AddInt64(&p.skips, 1)
 		return make([]byte, size)
 	}
-	var i int
-	for i = range BlockSizes {
-		if size <= BlockSizes[i] {
-			break
-		}
-	}
-	var bs []byte
+
 	// Try the fitting and all bigger pools
-	for j := i; j < len(BlockSizes); j++ {
+	var bs []byte
+	bkt := getBucketForSize(size)
+	for j := bkt; j < len(BlockSizes); j++ {
 		if intf := p.pools[j].Get(); intf != nil {
 			bs = *intf.(*[]byte)
+			atomic.AddInt64(&p.hits[j], 1)
 			return bs[:size]
 		}
 	}
-	// All pools are empty, must allocate.
-	return make([]byte, BlockSizes[i])[:size]
+
+	atomic.AddInt64(&p.misses, 1)
+
+	// All pools are empty, must allocate. For very small slices where we
+	// didn't have a block to reuse, just allocate a small slice instead of
+	// a large one. We won't be able to reuse it, but avoid some overhead.
+	if size < MinBlockSize/64 {
```

Comment:
Is this the mentioned "arena-limit"? If so, a comment about that would make the number less magic.

Reply:
It isn't, it's just something I picked out of thin air. I don't know what the specific limit is and it's anyway subject to change. If this is a contentious point I can remove the optimization...

Reply:
Ok fine, my point was just about mentioning it in comments, in case it was based on some arena thingy. Whether or not to have the optimization I don't have anything to base an opinion on, so go with whatever your gut says ;)